File: clipsync-win/ClipSync.Core/Settings/AppIdentity.cs
using System;

namespace ClipSync.Settings;

public enum AppKind
{
    /// Desktop (Win32) application, keyed by executable file name.
    Exe,
    /// Packaged (Store/UWP) application, keyed by package family name.
    Package,
}

/// Names an application for exclusion purposes.
///
/// Equality is deliberately on Kind+Key only: DisplayName and Path are
/// presentation data, so an app that is renamed or moves to a new
/// versioned install directory still matches a saved exclusion.
public sealed record AppIdentity
{
    public AppKind Kind { get; }

    /// Normalised match key. Lowercased throughout; for Exe this is the
    /// bare file name, so `app-1.0.9\Discord.exe` and
    /// `app-1.0.10\Discord.exe` are the same app.
    public string Key { get; }

    public string DisplayName { get; }

    /// Full path the user originally picked, kept only so the UI can show
    /// which `javaw.exe` (or similar ambiguous name) this refers to.
    public string? Path { get; }

    public AppIdentity(AppKind kind, string key, string displayName, string? path = null)
    {
        if (string.IsNullOrWhiteSpace(key)) throw new ArgumentException("key is required", nameof(key));
        Kind = kind;
        Key = Normalise(kind, key);
        DisplayName = displayName;
        Path = path;
    }

    private static string Normalise(AppKind kind, string key) => kind switch
    {
        AppKind.Exe => System.IO.Path.GetFileName(key.Trim()).ToLowerInvariant(),
        _ => key.Trim().ToLowerInvariant(),
    };

    public bool Equals(AppIdentity? other) =>
        other is not null && Kind == other.Kind && Key == other.Key;

    public override int GetHashCode() => HashCode.Combine(Kind, Key);
}

File: clipsync-mac/Sources/ClipSync/Settings/AppIdentity.swift
import Foundation

/// Names an application for exclusion purposes.
///
/// Equality is deliberately on kind+key only: displayName and path are
/// presentation data, so a renamed or relocated app still matches a saved
/// exclusion. Mirrors the Windows AppIdentity, where the same rule holds
/// for exe names and package family names.
struct AppIdentity: Equatable, Hashable {
    enum Kind: String {
        /// macOS application, keyed by bundle identifier. The only kind
        /// this platform produces; `exe`/`package` entries in a shared
        /// settings file are Windows-owned and ignored here.
        case bundle
    }

    let kind: Kind
    /// Normalised match key — the bundle identifier, trimmed and
    /// lowercased. The picker and the foreground resolver must both go
    /// through `normalise` so the stored key always equals the key the
    /// matcher produces (handoff §2.2).
    let key: String
    let displayName: String
    /// Path of the .app the user originally picked; presentation only.
    let path: String?

    init?(kind: Kind, key: String, displayName: String, path: String? = nil) {
        guard let normalised = AppIdentity.normalise(key) else { return nil }
        self.kind = kind
        self.key = normalised
        self.displayName = displayName
        self.path = path
    }

    /// One normalisation for both the picker and the resolver, by
    /// construction — bundle identifiers compare case-insensitively.
    static func normalise(_ key: String) -> String? {
        let trimmed = key.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return trimmed.isEmpty ? nil : trimmed
    }

    static func == (lhs: AppIdentity, rhs: AppIdentity) -> Bool {
        lhs.kind == rhs.kind && lhs.key == rhs.key
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(kind)
        hasher.combine(key)
    }
}
